modules/gatk_rnaseq/1.0/gatk_rnaseq.smk

```diff
@@ -0,0 +1,319 @@
+localrules:
+    _gatk_rnaseq_input_bam,
+    _gatk_rnaseq_output_vcf,
+    _gatk_rnaseq_all,
+
+
+##### RULES #####
+
+
+# Symlinks the input files into the module results directory (under '00-inputs/')
+rule _gatk_rnaseq_input_bam:
+    input:
+        bam = CFG["inputs"]["sample_bam"],
+        bai = CFG["inputs"]["sample_bai"]
+    output:
+        bam = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bam",
+        bai = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bai"
+    run:
+        op.absolute_symlink(input.bam, output.bam)
+        op.absolute_symlink(input.bai, output.bai)
+
+rule _gatk_splitntrim:
+    input:
+        bam = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bam",
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa")
+    output:
+        bam = temp(CFG["dirs"]["gatk_splitntrim"] +  "bam/{seq_type}--{genome_build}--{pair_status}/{sample_id}.split_reassign_mq.bam")
+    log:
+        stdout = CFG["logs"]["gatk_splitntrim"] + "{seq_type}--{genome_build}--{pair_status}/{sample_id}.gatk_splitntrim.stdout.log",
+        stderr = CFG["logs"]["gatk_splitntrim"] + "{seq_type}--{genome_build}--{pair_status}/{sample_id}.gatk_splitntrim.stderr.log"
+    params:
+        mem_mb = lambda wildcards, resources: int(resources.mem_mb / 1000)
+    conda:
+        CFG["conda_envs"]["gatk_rnaseq"]
+    threads:
+        CFG["threads"]["gatk_splitntrim"]
+    resources:
+        **CFG["resources"]["gatk_splitntrim"]
+    shell:
+        op.as_one_line("""
+        gatk --java-options "-Xmx{params.mem_mb}G" SplitNCigarReads -fixNDN TRUE -RF GoodCigarReadFilter
+        -R {input.fasta} -I {input.bam} -O {output.bam}
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+rule _gatk_base_recalibration:
+    input:
+        bam = rules._gatk_splitntrim.output,
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa")
+    output:
+        table = CFG["dirs"]["base_recal_report"] + "{seq_type}--{genome_build}--{pair_status}/{sample_id}.recalibration_report.grp"
+    log:
+        stdout = CFG["logs"]["base_recal_report"] + "{seq_type}--{genome_build}--{pair_status}/{sample_id}.gatk_base_recal.stdout.log",
+        stderr = CFG["logs"]["base_recal_report"] + "{seq_type}--{genome_build}--{pair_status}/{sample_id}.gatk_base_recal.stderr.log"
+    params:
+        dbsnp = reference_files("genomes/{genome_build}/variation/dbsnp.common_all-151.vcf.gz"),
+        gnomad = reference_files("genomes/{genome_build}/variation/af-only-gnomad.{genome_build}.vcf.gz"),
+        mem_mb = lambda wildcards, resources: int(resources.mem_mb / 1000)
+    conda:
+        CFG["conda_envs"]["gatk_rnaseq"]
+    threads: CFG["threads"]["gatk_base_recalibration"]
+    resources:
+        **CFG["resources"]["gatk_base_recalibration"]
+    shell:
+        op.as_one_line("""
+        gatk --java-options "-Xmx{params.mem_mb}G" BaseRecalibrator 
```

Comment:
I wonder if you know whether any rule is using java garbage collector, which may by default use all available threads? If so, it may create a lot of competitive usage when launched for many samples simultaneously. One way to get around this is to provide `-XX:ConcGCThreads=1` option that basically limits java to only 1 thread for garbage collector. Please ignore this comment if there is no issue with garbage collector!

Reply:
Hm I'm not sure about any rules specifically. I do see GATK documentation about some functions sometimes using too many resources. I will add `-XX:ConcGCThreads=1` as a pre-emptive measure